File: 2017-06-27-321-easy-talking-clock/main.cpp
#include <iostream>
#include <string>
#include <vector>

static const std::vector<std::string> ones = {
	"oh",
	"one",
	"two",
	"three",
	"four",
	"five",
	"six",
	"seven",
	"eight",
	"nine"
};

static const std::vector<std::string> tweens_teens = {
	"ten",
	"eleven",
	"twelve",
	"thirteen",
	"fourteen",
	"fifteen",
	"sixteen",
	"seventeen",
	"eighteen",
	"nineteen"
};

static const std::vector<std::string> tens = {
	"twenty",
	"thirty",
	"forty",
	"fifty"
};

std::string make_words(int n, bool is_hours)
{
	if (n == 0) return is_hours ? "twelve" : "";
	else if (n < 10) return (is_hours ? "" : "oh ") + ones.at(n);
	else if (n < 20) return tweens_teens.at(n % 10);
	else return tens.at(n / 10 - 2) + (n % 10 == 0 ? "" : " " + ones.at(n % 10));
}

int main()
{
	std::string line;
	while (std::getline(std::cin, line)) {
		auto colon_pos = line.find(':');
		int hours = std::stoi(line.substr(0, colon_pos));
		int minutes = std::stoi(line.substr(colon_pos + 1));
		bool past_noon = (hours >= 12 ? true : false);
		hours = (hours >= 12 ? hours == 12 ? 12 : hours % 12 : hours);
		auto hours_str = make_words(hours, true);
		auto mins_str = make_words(minutes, false);
		std::cout << "It's " << hours_str << " " << mins_str;
		if (!mins_str.empty()) std::cout << " ";
		std::cout << (past_noon ? "pm" : "am") << '\n';
	}
	return 0;
}

File: 2017-06-27-321-easy-talking-clock/main.swift
import Foundation

func makeTimeWords(forTime time: Int, isHours: Bool) -> String {
    let ones = ["oh", "one", "two", "three", "four", "five", "six",
        "seven", "eight", "nine"]
    let tweensTeens = ["ten", "eleven", "twelve", "thirteen", "fourteen",
        "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"]
    let tens = ["twenty", "thirty", "forty", "fifty"]
    if time == 0 { return isHours ? "twelve" : "" }
    else if time < 10 { return "\(isHours ? "" : "oh ")\(ones[time])" }
    else if time < 20 { return tweensTeens[time % 10] }
    else { return tens[time / 10 - 2] + (time % 10 == 0 ? "" : " \(ones[time % 10])") }
}

func speakTime(forTime input: String) -> String {
    let colonIdx = input.firstIndex(of: ":") ?? input.endIndex
    var hours = Int(String(input[input.startIndex..<colonIdx])) ?? 0
    let afterColonIdx = input.index(after: colonIdx)
    let minutes = Int(String(input[afterColonIdx...])) ?? 0
    let pastNoon = hours >= 12
    hours = (hours >= 12 ? hours == 12 ? 12 : hours % 12 : hours)
    let hoursStr = makeTimeWords(forTime: hours, isHours: true)
    let minsStr = makeTimeWords(forTime: minutes, isHours: false)
    return "It's \(hoursStr) \(minsStr)\(!minsStr.isEmpty ? " " : "")\(pastNoon ? "pm" : "am")"
}
